main/ts_client.c

```diff
@@ -5,67 +5,251 @@
  */
 
 #include "ts_client.h"
+#include "ts_serial.h"
 
 #include <stdio.h>
 #include <stdlib.h>
 
 #include "esp_http_server.h"
 #include "esp_err.h"
 #include "esp_log.h"
+#include "cJSON.h"
 
 static const char *TAG = "ts_client";
+static TSDevice *devices[10];
 
-char *ts_resp_data(char *resp)
+
+//scan devices and initialise device list with NULL
+void ts_scan_devices()
 {
-    if (resp[0] == ':') {
-        char *pos = strstr(resp, ". ");
-        if (pos != NULL) {
-            return pos + 2;
+    for (int i = 0; i < 10; i++) {
+        devices[i] = NULL;
+    }
+    //scan serial connection
+    devices[0] = (TSDevice *) heap_caps_malloc(sizeof(TSDevice), MALLOC_CAP_8BIT);
+    if (ts_serial_scan_device_info(devices[0]) != 0) {
+        free(devices[0]);
+        devices[0] = NULL;
+    };
+
+    //TODO scan CAN connection
+}
+
+char *ts_get_device_list()
+{
+    cJSON *names = cJSON_CreateArray();
+    int i = 0;
+    while (devices[i] != NULL) {
+        cJSON *name = cJSON_CreateString(devices[i]->ts_name);
+        cJSON_AddItemToArray(names, name);
+        i++;
+    }
+    char *names_string = NULL;
+    if (i > 0) {
+        names_string = cJSON_Print(names);
+        cJSON_Delete(names);
+    } else {
+        // hackky solution so that free() can always be called on names_string
+        char *msg = "No devices Connected";
+        names_string = (char *) malloc(strlen(msg)+1*sizeof(char));
+        strncpy(names_string, msg, strlen(msg)+1);
```

Comment:
Some small optimizations: Declaring the string const saves some memory in RAM. And I think we could use `sizeof` (which is evaluated at compile-time) instead of `strlen`.

```suggestion
        const char msg[] = "No devices Connected";
        names_string = (char *)malloc(sizeof(msg));
        strncpy(names_string, msg, sizeof(msg));
```